### ISSUE TITLE
simplifyDefUse does not remove redundant assignments to slices 

### ISSUE THREAD

**Issue by jphurl** [p4lang/p4c#2583]
The following assignments are reduced by `simplifyDefUse` pass:
```p4
h.data = 8;
h.data = 8;
```
to just:
```p4
h.data = 8;
```
but the same is not true if the lhs is a slice:
```p4
h.data[7:0] = 8;
h.data[7:0] = 8;
```
this does not get simplified by `simplifyDefUse`

here is a test: 
[simplify-slice.p4.txt](https://github.com/p4lang/p4c/files/5373502/simplify-slice.p4.txt)

**Comment by mbudiu-vmw** [comment]
Yes, our def-use analysis is not precise enough to track bits, only whole values. Would be nice to do this someday.

**Comment by dbrajovic-syrmia** [comment]
> Yes, our def-use analysis is not precise enough to track bits, only whole values. Would be nice to do this someday.

If you approve, I would like this issue assigned to me.

**Comment by mbudiu-vmw** [comment]
We value any contributions. However, I have to warn you that I am not particularly proud of the def-use pass implementation. I find it difficult to debug myself. There are in fact two passes: def-use computes reaching definitions, and simplifyDefUse uses this information to remove dead assignments. The interplay between them isn't very nice. Perhaps our IR is too high-level? Modelling the semantics of some operations, like method calls with out arguments, is quite tricky. The pass needs to be interprocedural, because some "methods" (e.g., actions) can have side-effects and can access variables in the environment that are not passed as arguments.

**Comment by mbudiu-vmw** [comment]
Note that this pass runs after the SideEFfectOrdering pass, which simplifies some even trickier code cases into simpler ones (e.g., method calls where an argument is another method call that can have side-effects).

**Comment by dbrajovic-syrmia** [comment]
>There are in fact two passes: def-use computes reaching definitions, and simplifyDefUse uses this information to remove dead assignments.

Speaking of SimplifyDefUse and its decision-making to remove dead statements; it seems that for non-slice statements (such as the above `h.data = 8`), the `getReads()` method returns an empty `LocationSet` (in `registerUses()`) which is why the the pass is able to remove the duplicate statement further down the stack. The duplicate doesn't make it to the `HasUses::used` map and for that reason it gets thrown out.

However, with slices we have a different flow, mostly thanks to the `bool FindUninitialized::lhs` member variable. Inside the `bool preorder(const IR::Slice* expression) override` method, this variable gets intentionally set to false, which causes the `if` clause inside of `bool preorder(const IR::PathExpression* expression) override` to be skipped and not return an empty `LocationSet` as a result (unlike our non-slice example).

So far this is just an observation - I have yet to explore this logic and its underlying design. 

But, I did stumble upon a rather misconstrued piece of code - it's the `HasUses::add()` method.
(https://github.com/p4lang/p4c/blob/c1cf3c4a5dcad6f9ee846fda96e65e18c0863f77/frontends/p4/simplifyDefUse.cpp#L31)

The original intent definitely seems to be keeping track of nodes and their use count by filling in the `HasUses::used` map. But if we print out the actual values inside it, we'll see that all nodes have a value of 1 regardless of their "increment" in the else clause. This is because the `emplace` inside the clause will only happen if the key being inserted is unique. Since this can't be the case (because we're in the `else` branch), the increment never takes place for any of the nodes.

**Comment by mbudiu-vmw** [comment]
Indeed, initially I thought I could use the same map while removing unused assignments from the program, but then I realized that I still need to recompute the def-use after removing an assignment, and thus I cannot use this map incrementally, so I gave up maintaining a count. Only used/unused matters currently.
